### PR TITLE
restore jobs with file

### PR DESCRIPTION
**Pull request by vision9527**
restore jobs with file

### CODE REVIEW

**Comment by vision9527** [comment]
@victorcoder 
I have removed download api and squash to one commit.

**Comment by victorcoder** [comment]
Thanks for your work @vision9527 are you still interested in this?

**Comment by vision9527** [comment]
> Thanks for your work @vision9527 are you still interested in this?

I'm just in this case. Respect your design philosophy. It's up to you.

File: dkron/api.go [hunk 1]
```diff
@@ -73,6 +75,7 @@ func (h *HTTPTransport) APIRoutes(r *gin.RouterGroup, middleware ...gin.HandlerF
 	v1.PATCH("/jobs", h.jobCreateOrUpdateHandler)
 	// Place fallback routes last
 	v1.GET("/jobs", h.jobsHandler)
+	v1.POST("/restoreJobs", h.restoreJobsHandler)
```
Comment (victorcoder) [inline dkron/api.go]: I would call this just `/restore` or `/load`

File: dkron/api.go [hunk 2]
```diff
@@ -212,6 +215,42 @@ func (h *HTTPTransport) jobRunHandler(c *gin.Context) {
 	renderJSON(c, http.StatusOK, job)
 }
 
+// Restore jobs from file.
+// Overwrite job if the job is exist.
+func (h *HTTPTransport) restoreJobsHandler(c *gin.Context) {
```
Comment (victorcoder) [inline dkron/api.go]: same for handler `restoreHandler`

**Comment by victorcoder** [comment]
Also modify the `swagger.yml` spec please.

**Comment by victorcoder** [comment]
@vision9527 also you will need to fix conflicts

File: dkron/api.go (hunk 2 above)
Comment (vision9527) [reply dkron/api.go]: done

**Comment by vision9527** [comment]
@victorcoder done

File: dkron/job.go [hunk 3]
```diff
@@ -396,3 +399,68 @@ func isSlug(candidate string) (bool, string) {
...
+		}
+		if isTopParentNodeFlag {
+			j++
+		}
+		jobs = rejobs
+	}
+	return jobs, nil
+}
+
+// find parent job and validate job
```
Comment (victorcoder) [inline dkron/job.go]: This is not correctly linted, it should start with `findParentJobAndValidateJob...`

**Comment by victorcoder** [comment]
@vision9527 could you write the rationale for creating a job tree. I see some test jobs without dependent jobs. In theory if the jobs were exported with the `/jobs` endpoint, importing them one by one should end up with a correct structure without having to elaborate such a complex parsing.

**Comment by vision9527** [comment]
> @vision9527 could you write the rationale for creating a job tree. I see some test jobs without dependent jobs. In theory if the jobs were exported with the `/jobs` endpoint, importing them one by one should end up with a correct structure without having to elaborate such a complex parsing.

Oh, it's my mistake. I change the test jobs code. You can review now.  The job tree that has parent job and dependent job can be created with a correct structure. The key is to deal with parent job and dependent job in the endpoint.

**Comment by victorcoder** [comment]
@vision9527 Shouldn't the enpoint in this comment https://github.com/distribworks/dkron/pull/654#discussion_r393615297 be just `restore`

I will ask it in another way, what will happen if we just do:

```go
// file: api.go
...
var jobs []*Job
err = json.Unmarshal(data, &jobs)

for _, job := range jobs {
  err := a.GRPCClient.SetJob(job)
}
...
```

**Comment by vision9527** [comment]
> @vision9527 Shouldn't the enpoint in this comment [#654 (comment)](https://github.com/distribworks/dkron/pull/654#discussion_r393615297) be just `restore`
> 
> I will ask it in another way, what will happen if we just do:
> 
> ```go
> // file: api.go
> ...
> var jobs []*Job
> err = json.Unmarshal(data, &jobs)
> 
> for _, job := range jobs {
>   err := a.GRPCClient.SetJob(job)
> }
> ...
> ```

image
Abort if parent not found before committing job to the store.
It can't be for sure that the create job sequence is right.
Job tree is a array in which parent job is in the front of  the dependent jobs, so it can be sure when  the dependent job will be created that parent job is exist.

**Comment by vision9527** [comment]
@victorcoder Please review, bro.

**Comment by victorcoder** [comment]
Many changes in master, please fix conflicts or rebase, and we should be good to go, thanks!

**Comment by vision9527** [comment]
> Many changes in master, please fix conflicts or rebase, and we should be good to go, thanks!

Done.

File: dkron/store.go [hunk 4]
```diff
@@ -149,6 +149,14 @@ func (s *Store) SetJob(job *Job, copyDependentJobs bool) error {
 		if err := s.removeFromParent(ej); err != nil {
 			return err
 		}
+
+		if ej == nil || ej.ParentJob == "" {
```
Comment (victorcoder) [inline dkron/store.go:153]: ej can't be nil at this point, no need to check this

File: dkron/store.go [hunk 5]
```diff
@@ -149,6 +149,14 @@ func (s *Store) SetJob(job *Job, copyDependentJobs bool) error {
...
+			// Need to handle this case already created parent job when dependent job is not exist
```
Comment (victorcoder) [inline dkron/store.go:154]: Please reword this comment as it's not clear what it means.

File: dkron/store.go (hunk 4 above)
Comment (vision9527) [reply dkron/store.go]: yes, you are right. I have clean this.

**Pull request merged**